src/icu_win.cc

```diff
@@ -0,0 +1,238 @@
+// wchar_t and char16_t are equivalent on Windows.
+using UChar = wchar_t;
+
+enum UErrorCode : std::int32_t {
+  U_ZERO_ERROR = 0,
+  U_BUFFER_OVERFLOW_ERROR = 15,
+};
+
+bool U_SUCCESS(UErrorCode error) { return error <= U_ZERO_ERROR; }
+
+// ICU function signatures
+using ucal_getHostTimeZone_func = std::int32_t(__cdecl*)(
+    UChar* result, std::int32_t result_capacity, UErrorCode* status);
+using ucal_getTimeZoneIDForWindowsID_func = std::int32_t(__cdecl*)(
+    const UChar* winid, std::int32_t len, const char* region, UChar* id,
+    std::int32_t id_capacity, UErrorCode* status);
+
+std::atomic<bool> g_unavailable;
+std::atomic<ucal_getHostTimeZone_func> g_ucal_getHostTimeZone;
+std::atomic<ucal_getTimeZoneIDForWindowsID_func>
+    g_ucal_getTimeZoneIDForWindowsID;
+
+struct IcuFunctions {
+  const bool available;
+  const ucal_getHostTimeZone_func ucal_getHostTimeZone;
+  const ucal_getTimeZoneIDForWindowsID_func ucal_getTimeZoneIDForWindowsID;
+};
+
+IcuFunctions Unavailable() { return {false, nullptr, nullptr}; }
+
+template <typename T>
+static T AsProcAddress(HMODULE module, const char* name) {
+  static_assert(
+      std::is_pointer<T>::value &&
+          std::is_function<typename std::remove_pointer<T>::type>::value,
+      "T must be a function pointer type");
+  const auto proc_address = ::GetProcAddress(module, name);
+  return reinterpret_cast<T>(reinterpret_cast<void*>(proc_address));
+}
+
+std::wstring GetSystem32Dir() {
+  size_t dynamic_buffer_size = 0;
+  {
+    // Fast-path. Usually it's "C:\Windows\System32", which is 19 characters.
+    std::array<wchar_t, 32> buffer;
+    UINT result = ::GetSystemDirectoryW(buffer.data(), buffer.size());
+    // If successful, the return value is the length of the string without NUL.
+    if (result <= buffer.size()) {
+      return std::wstring(buffer.data(), result);
+    }
+    // Otherwise, the return value is the required length + 1 for NUL.
+    dynamic_buffer_size = result;
+  }
+  // Slow-path. Allocate a buffer of the required size.
+  std::unique_ptr<wchar_t[]> buffer(new wchar_t[dynamic_buffer_size]);
+  const UINT length = ::GetSystemDirectoryW(buffer.get(), dynamic_buffer_size);
+  if (length <= dynamic_buffer_size) {
+    return std::wstring(buffer.get(), length);
+  }
+  return std::wstring();
+}
+
+IcuFunctions GetIcuFunctions() {
+  if (g_unavailable.load(std::memory_order_relaxed)) {
+    return Unavailable();
+  }
+
+  // Check if already loaded
+  {
+    const auto ucal_getHostTimeZoneRef =
+        g_ucal_getHostTimeZone.load(std::memory_order_relaxed);
+    const auto ucal_getTimeZoneIDForWindowsIDRef =
+        g_ucal_getTimeZoneIDForWindowsID.load(std::memory_order_relaxed);
+
+    // Note: ucal_getHostTimeZone can be unavailable on older Windows.
+    if (ucal_getTimeZoneIDForWindowsIDRef != nullptr) {
+      return {true, ucal_getHostTimeZoneRef, ucal_getTimeZoneIDForWindowsIDRef};
+    }
+  }
+
```

Comment:
Two threads could be here at the same time.  Is that a bug, a feature, or a don't-care?

Reply:
This is intended to be lock-free so this is a kind of feature. Added a comment at the beginning of the function.